File: http/http-client-annotation-processor/src/main/java/ru/tinkoff/kora/http/client/annotation/processor/ConfigModuleGenerator.java
package ru.tinkoff.kora.http.client.annotation.processor;

import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.common.Module;
import ru.tinkoff.kora.common.annotation.Generated;
import ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor;
import ru.tinkoff.kora.http.client.common.annotation.HttpClient;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import java.util.Optional;

public class ConfigModuleGenerator {
    private final Elements elements;

    public ConfigModuleGenerator(ProcessingEnvironment processingEnvironment) {
        this.elements = processingEnvironment.getElementUtils();
    }

    public JavaFile generate(TypeElement element) {
        var lowercaseName = new StringBuilder(element.getSimpleName());
        lowercaseName.setCharAt(0, Character.toLowerCase(lowercaseName.charAt(0)));
        var packageName = this.elements.getPackageOf(element).getQualifiedName().toString();
        var configPath = element.getAnnotation(HttpClient.class).configPath();
        if (configPath.isBlank()) {
            configPath = "httpClient." + lowercaseName;
        }

        var configName = HttpClientUtils.configName(element);
        var moduleName = HttpClientUtils.moduleName(element);
        var configClass = ClassName.get(packageName, configName);
        var extractorClass = ParameterizedTypeName.get(ClassName.get(ConfigValueExtractor.class), configClass);

        var type = TypeSpec.interfaceBuilder(moduleName)
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(AnnotationSpec.builder(Generated.class).addMember("value", "$S", ConfigModuleGenerator.class.getCanonicalName()).build())
            .addAnnotation(AnnotationSpec.builder(Module.class).build())
            .addOriginatingElement(element)
            .addMethod(MethodSpec.methodBuilder(lowercaseName + "Config")
                .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                .returns(configClass)
                .addParameter(ParameterSpec.builder(CommonClassNames.config, "config").build())
                .addParameter(ParameterSpec.builder(extractorClass, "extractor").build())
                .addStatement("var value = config.get($S)", configPath)
                .addStatement("return $T.ofNullable(extractor.extract(value)).orElseThrow(() -> $T.missingValueAfterParse(value))", Optional.class, CommonClassNames.configValueExtractionException)
                .build());

        return JavaFile.builder(packageName, type.build()).build();
    }
}

File: http/http-client-symbol-processor/src/main/kotlin/ru/tinkoff/kora/http/client/symbol/processor/ConfigModuleGenerator.kt
package ru.tinkoff.kora.http.client.symbol.processor

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.KspCommonUtils.generated

class ConfigModuleGenerator(val resolver: Resolver) {

    @KspExperimental
    fun generate(declaration: KSClassDeclaration): FileSpec {
        val lowercaseName = StringBuilder(declaration.simpleName.asString())
        lowercaseName.setCharAt(0, lowercaseName[0].lowercaseChar())
        val packageName = declaration.packageName.asString()
        var configPath = declaration.findAnnotation(ClassName("ru.tinkoff.kora.http.client.common.annotation", "HttpClient"))
            ?.findValue<String>("configPath")!!
        if (configPath.isBlank()) {
            configPath = "httpClient.$lowercaseName"
        }
        val configName = declaration.configName()
        val moduleName = declaration.moduleName()
        val configClass = ClassName(packageName, configName)
        val extractorClass = CommonClassNames.configValueExtractor.parameterizedBy(configClass)
        val type = TypeSpec.interfaceBuilder(moduleName)
            .generated(ConfigModuleGenerator::class)
            .addAnnotation(AnnotationSpec.builder(CommonClassNames.module).build())
            .addOriginatingKSFile(declaration.containingFile!!)
            .addFunction(
                FunSpec.builder(lowercaseName.toString() + "Config")
                    .returns(configClass)
                    .addParameter(ParameterSpec.builder("config", CommonClassNames.config).build())
                    .addParameter(ParameterSpec.builder("extractor", extractorClass).build())
                    .addStatement("val value = config.get(%S)", configPath)
                    .addStatement("return extractor.extract(value) ?: throw %T.missingValueAfterParse(value)", CommonClassNames.configParseException)
                    .build()
            )
        return FileSpec.builder(packageName, moduleName)
            .addType(type.build())
            .build()
    }
}
